### PR TITLE
Additions for Ethereals.

### PR DESCRIPTION
**Pull request by dekunutkid**
## About The Pull Request
This is a port from TG that will allow Ethereals to charge from APCs, deliver their own power to APCs in order to keep them operational, as well as being able to charge from power cells, albeit far less efficiently. As well as give them more customization options, rather than just color.
(Sorry for the weird artefacts in the commits that I couldn't figure out how to remove)

## Why It's Good For The Game
These changes should help make Ethereals a more popular choice, instead of just "a weird organic IPC" due to their newfound utility and self-sustenance. being still relatively frail, but also not having to rely as much on help to keep going.

## Changelog
:cl: 
add: The ability for Ethereals to consume power from APCs while on harm intent, and give APCs power while on grab intent.
add: All hair options are available for Ethereals with the color matching the Ethereal's color.
add: Ethereals can now use a power cell in their hand and drain it at a 5% nutrition yield.
/:cl:

### CODE REVIEW

File: .vs/VSWorkspaceState.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (ike709) [inline .vs/VSWorkspaceState.json]: ...This file shouldn't be here...

**Comment by ike709** [comment]
That sqlite file also shouldn't be there... How'd you manage that?

**Comment by Lagomorphica** [comment]
cool and good

**Comment by dekunutkid** [comment]
like i said, weird artefacts and i have no clue how to remove them, or how they even got to commit in the first place.

File: code/modules/power/apc.dm [hunk 2]
```diff
@@ -799,6 +799,39 @@
 	. = ..()
 	if(.)
 		return
+	var/mob/living/carbon/human/H = user
+	var/datum/species/ethereal/eth_species = H.dna?.species
+	if(istype(H) && istype(eth_species))
+		if(H.a_intent == INTENT_HARM)
+			if(eth_species.ethereal_charge > 145)
```
Comment (ike709) [inline code/modules/power/apc.dm]:
```suggestion
			if(eth_species.ethereal_charge >= ETHERAL_CHARGE_FULL)
```

File: code/modules/power/apc.dm [hunk 3]
```diff
@@ -799,6 +799,39 @@
...
+				to_chat(H, "<span class='warning'>Your charge is full!</span>")
+				return
+			to_chat(H, "<span class='notice'>You start channeling some power through the APC into your body.</span>")
```
Comment (ike709) [inline code/modules/power/apc.dm:812]: You need to check that the cell actually has charge before this.

Comment (ike709) [reply code/modules/power/apc.dm:812]: Like that `cell.charge >= (cell.maxcharge / 2)` bit should be up here instead.

File: code/modules/power/apc.dm [hunk 4]
```diff
@@ -799,6 +799,39 @@
...
+			if(do_after(user, 70, target = src))
+				if(cell.charge >= (cell.maxcharge / 2) || (eth_species.ethereal_charge >= ETHEREAL_CHARGE_FULL))
```
Comment (ike709) [inline code/modules/power/apc.dm]:
```suggestion
				if(cell.charge >= 100 && eth_species.ethereal_charge < ETHEREAL_CHARGE_FULL)
```

File: code/modules/power/apc.dm [hunk 5]
```diff
@@ -799,6 +799,39 @@
...
+					to_chat(H, "<span class='notice'>You receive some charge from the APC.</span>")
+					eth_species.adjust_charge(20)
+					cell.charge -= 20
```
Comment (ike709) [inline code/modules/power/apc.dm]:
```suggestion
					cell.charge -= 50 //Charging is inefficient
```

File: code/modules/power/apc.dm [hunk 6]
```diff
@@ -799,6 +799,39 @@
...
+	if(istype(H) && istype(eth_species))
+		if(H.a_intent == INTENT_HARM)
+			if(eth_species.ethereal_charge > 145)
+				to_chat(H, "<span class='warning'>Your charge is full!</span>")
+				return
+			to_chat(H, "<span class='notice'>You start channeling some power through the APC into your body.</span>")
+			if(do_after(user, 70, target = src))
+				if(cell.charge >= (cell.maxcharge / 2) || (eth_species.ethereal_charge >= ETHEREAL_CHARGE_FULL))
+					to_chat(H, "<span class='notice'>You receive some charge from the APC.</span>")
+					eth_species.adjust_charge(20)
```
Comment (ike709) [inline code/modules/power/apc.dm:816]: I don't know if 20 is a lot or not. Is 20 a lot? what's the max charge on ethereals?

File: code/modules/power/apc.dm [hunk 7]
```diff
@@ -799,6 +799,39 @@
...
+					cell.charge -= 20
+				else
+					to_chat(H, "<span class='warning'>You can't receive charge from the APC!</span>")
```
Comment (ike709) [inline code/modules/power/apc.dm:816]:
```suggestion
					to_chat(H, "<span class='warning'>You fail to receive charge from the APC!</span>")
```

File: code/modules/power/apc.dm [hunk 8]
```diff
@@ -799,6 +799,39 @@
...
+			return
+		if(H.a_intent == INTENT_GRAB)
+			if(cell.charge == cell.maxcharge)
```
Comment (ike709) [inline code/modules/power/apc.dm]:
```suggestion
			if(cell.charge >= cell.maxcharge - 20)
```

File: code/modules/power/apc.dm [hunk 9]
```diff
@@ -799,6 +799,39 @@
...
+				to_chat(H, "<span class='warning'>The APC is full!</span>")
+				return
+			if(eth_species.ethereal_charge < 10)
```
Comment (ike709) [inline code/modules/power/apc.dm]:
```suggestion
			if(eth_species.ethereal_charge < 20)
```

File: code/modules/power/apc.dm (hunk 6 above)
Comment (dekunutkid) [reply code/modules/power/apc.dm:816]: max charge on ethereals is ~150

File: code/modules/power/apc.dm [hunk 10]
```diff
@@ -799,6 +799,39 @@
...
+				to_chat(H, "<span class='warning'>Your charge is too low!</span>")
+				return
+			to_chat(H, "<span class='notice'>You start channeling power through your body into the APC.</span>")
+			if(do_after(user, 70, target = src))
+				if(cell.charge >= cell.maxcharge || (eth_species.ethereal_charge > 10))
```
Comment (ike709) [inline code/modules/power/apc.dm]:
```suggestion
				if(cell.charge <= cell.maxcharge + 20 && (eth_species.ethereal_charge > 20))
```

File: code/modules/power/apc.dm [hunk 11]
```diff
@@ -799,6 +799,39 @@
...
+			return
+		if(H.a_intent == INTENT_GRAB)
+			if(cell.charge == cell.maxcharge)
+				to_chat(H, "<span class='warning'>The APC is full!</span>")
+				return
+			if(eth_species.ethereal_charge < 10)
+				to_chat(H, "<span class='warning'>Your charge is too low!</span>")
+				return
+			to_chat(H, "<span class='notice'>You start channeling power through your body into the APC.</span>")
+			if(do_after(user, 70, target = src))
```
Comment (ike709) [inline code/modules/power/apc.dm]:
```suggestion
			if(do_after(user, 30, target = src))
```

File: code/modules/power/apc.dm [hunk 12]
```diff
@@ -799,6 +799,39 @@
...
 		return
+	var/mob/living/carbon/human/H = user
+	var/datum/species/ethereal/eth_species = H.dna?.species
+	if(istype(H) && istype(eth_species))
+		if(H.a_intent == INTENT_HARM)
+			if(eth_species.ethereal_charge > 145)
+				to_chat(H, "<span class='warning'>Your charge is full!</span>")
+				return
+			to_chat(H, "<span class='notice'>You start channeling some power through the APC into your body.</span>")
+			if(do_after(user, 70, target = src))
```
Comment (ike709) [inline code/modules/power/apc.dm]:
```suggestion
			if(do_after(user, 30, target = src))
```

File: code/modules/power/cell.dm [hunk 13]
```diff
@@ -148,6 +148,27 @@
...
 
+/obj/item/stock_parts/cell/attack_self(mob/user)
+	var/mob/living/carbon/human/H = user
+	var/datum/species/ethereal/eth_species = H.dna?.species
+	
+	if(istype(H) && istype(eth_species))			
+		if(charge < 100)
+			to_chat(H, "<span class='warning'>The [src] doesn't have enough power!</span>")
+			return
+		if(eth_species.ethereal_charge == ETHEREAL_CHARGE_FULL)
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion
		if(eth_species.ethereal_charge >= ETHEREAL_CHARGE_FULL - 5)

```

File: code/modules/power/cell.dm [hunk 14]
```diff
@@ -148,6 +148,27 @@
...
+			to_chat(H, "<span class='warning'>Your charge is full!</span>")
+			return
+		to_chat(H, "<span class='notice'>You clumsily channel power through the [src] and into your body, wasting some in the process.</span>")
+		if(do_after(user, 5, target = src))
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion
		if(do_after(user, 50, target = src))

```

File: code/modules/power/cell.dm [hunk 15]
```diff
@@ -148,6 +148,27 @@
...
+			if((charge >= 100) && (eth_species.ethereal_charge < ETHEREAL_CHARGE_FULL))
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion
			if((charge >= 100) && (eth_species.ethereal_charge < ETHEREAL_CHARGE_FULL - 5))

```

File: code/modules/surgery/bodyparts/head.dm [hunk 16]
```diff
@@ -148,6 +148,8 @@
 			if(S.hair_color)
 				if(S.hair_color == "mutcolor")
 					facial_hair_color = H.dna.features["mcolor"]
+				else if(hair_color == "fixedmutcolor")
```
Comment (ike709) [inline code/modules/surgery/bodyparts/head.dm]:
```suggestion
				else if(S.hair_color == "fixedmutcolor")
```

File: code/modules/surgery/bodyparts/head.dm [hunk 17]
```diff
@@ -163,6 +165,8 @@
 			if(S.hair_color)
 				if(S.hair_color == "mutcolor")
 					hair_color = H.dna.features["mcolor"]
+				else if(hair_color == "fixedmutcolor")
```
Comment (ike709) [inline code/modules/surgery/bodyparts/head.dm]:
```suggestion
				else if(S.hair_color == "fixedmutcolor")
```

File: code/modules/power/apc.dm [hunk 18]
```diff
@@ -799,6 +799,39 @@
 	. = ..()
 	if(.)
 		return
+	var/mob/living/carbon/human/H = user
+	var/datum/species/ethereal/eth_species = H.dna?.species
+	if(istype(H) && istype(eth_species))
+		if(H.a_intent == INTENT_HARM)
+			if(eth_species.ethereal_charge >= ETHERAL_CHARGE_FULL)
```
Comment (ike709) [inline code/modules/power/apc.dm]:
```suggestion
			if(eth_species.ethereal_charge >= ETHEREAL_CHARGE_FULL)
```

**Comment by qwertyquerty** [comment]
wew

File: code/modules/power/cell.dm [hunk 19]
```diff
@@ -148,6 +148,30 @@
...
+/obj/item/stock_parts/cell/attack_self(mob/user)
+	var/mob/living/carbon/human/H = user
+	var/datum/species/ethereal/eth_species = H.dna?.species
+	
+	if(istype(H) && istype(eth_species))			
+		if(charge < 100)
+			to_chat(H, "<span class='warning'>The [src] doesn't have enough power!</span>")
+			return
+		if(eth_species.ethereal_charge >= ETHEREAL_CHARGE_FULL - 5)
+
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion

```

File: code/modules/power/cell.dm [hunk 20]
```diff
@@ -148,6 +148,30 @@
...
+			to_chat(H, "<span class='warning'>Your charge is full!</span>")
+			return
+		to_chat(H, "<span class='notice'>You clumsily channel power through the [src] and into your body, wasting some in the process.</span>")
+		if(do_after(user, 50, target = src))
+
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion

```

File: code/modules/power/cell.dm [hunk 21]
```diff
@@ -148,6 +148,30 @@
...
+			if((charge >= 100) && (eth_species.ethereal_charge < ETHEREAL_CHARGE_FULL - 5))
+
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion

```

File: code/modules/power/cell.dm [hunk 22]
```diff
@@ -148,6 +148,30 @@
...
+				to_chat(H, "<span class='notice'>You receive some charge from the [src].</span>")
+				eth_species.adjust_charge(5)
+				charge -= 100 //you waste way more than you receive, so that ethereals can't just steal one cell and forget about hunger
+			else
+				to_chat(H, "<span class='warning'>You can't receive charge from the [src]!</span>")
+		return
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion
			return

```

File: code/modules/power/cell.dm [hunk 23]
```diff
@@ -148,6 +148,30 @@
...
+
```
Comment (ike709) [inline code/modules/power/cell.dm:172]:
```suggestion
		
..()
```

File: code/modules/power/cell.dm [hunk 24]
```diff
@@ -148,6 +148,29 @@
...
+			if((charge >= 100) && (eth_species.ethereal_charge < ETHEREAL_CHARGE_FULL - 5))
+				to_chat(H, "<span class='notice'>You receive some charge from the [src].</span>")
+				eth_species.adjust_charge(5)
+				charge -= 100 //you waste way more than you receive, so that ethereals can't just steal one cell and forget about hunger
+			else
+				to_chat(H, "<span class='warning'>You can't receive charge from the [src]!</span>")
+			return
+
+		
+..()
```
Comment (ike709) [inline code/modules/power/cell.dm]:
```suggestion
	..()

```

File: code/modules/power/apc.dm [hunk 25]
```diff
@@ -799,6 +799,39 @@
...
+		if(H.a_intent == INTENT_GRAB)
+			if(cell.charge >= cell.maxcharge)
+				to_chat(H, "<span class='warning'>The APC is full!</span>")
+				return
+			if(eth_species.ethereal_charge < 20)
+				to_chat(H, "<span class='warning'>Your charge is too low!</span>")
+				return
+			to_chat(H, "<span class='notice'>You start channeling power through your body into the APC.</span>")
+			if(do_after(user, 30, target = src))
+				if(cell.charge <= cell.maxcharge + 20 && (eth_species.ethereal_charge > 20))
```
Comment (ike709) [inline code/modules/power/apc.dm]:
```suggestion
				if(cell.charge + 20 <= cell.maxcharge && (eth_species.ethereal_charge > 20))
```

File: code/modules/power/apc.dm (hunk 3 above)
Comment (ike709) [reply code/modules/power/apc.dm:812]: You still need to check the cell charge before the `do_after`

File: code/modules/power/apc.dm [hunk 26]
```diff
@@ -803,12 +803,15 @@
 	var/datum/species/ethereal/eth_species = H.dna?.species
 	if(istype(H) && istype(eth_species))
 		if(H.a_intent == INTENT_HARM)
-			if(eth_species.ethereal_charge >= ETHEREAL_CHARGE_FULL)
+			if(eth_species.ethereal_charge >= ETHEREAL_CHARGE_FULL - 19)
 				to_chat(H, "<span class='warning'>Your charge is full!</span>")
 				return
+					if(cell.charge < 100)
```
Comment (ike709) [inline code/modules/power/apc.dm]: I told you to ignore discord formatting, this code is indented too far and should be aligned with the previous check.

**Review by ike709 — APPROVED** [review]

**Review by qwertyquerty — APPROVED** [review]

**Pull request merged**